usr/lib/ica_s390_stdll/ica_specific.c

```diff
@@ -3124,6 +3125,364 @@ static CK_RV os_specific_rsa_decrypt(STDLL_TokData_t *tokdata,
+                                    0, gcm_ctx);
+
+        /* save any remaining data */
+        if (remain != 0)
+            memcpy(context->data, in_data + (in_data_len - remain), remain);
+        context->len = remain;
+    } else {
+        /* decrypt */
+        /* copy all the leftover data from previous decryption first */
+        if (in_data_len >= tag_data_len) {      /* case 1  */
+            /* copy complete context to buffer first */
+            memcpy(buffer, context->data, context->len);
+            /* Append in_data to buffer */
+            memcpy(buffer + context->len, in_data, out_len - context->len);
+            /* copy remaining data to context */
+            memcpy(context->data, in_data + out_len - context->len, remain);
+            context->len = remain;
+        } else {                /* case 2 - partial data */
+            memcpy(buffer, context->data, AES_BLOCK_SIZE);
+            memmove(context->data, context->data + AES_BLOCK_SIZE,
+                    context->len - AES_BLOCK_SIZE);
+            memcpy(context->data + context->len - AES_BLOCK_SIZE,
+                   in_data, in_data_len);
+            context->len = context->len - AES_BLOCK_SIZE + in_data_len;
+        }
+
+        rc = ica_aes_gcm_kma_update(buffer, out_data, out_len,
+                                    auth_data, auth_data_len,
+                                    context->ulAlen ? 0 : 1,
+                                    0, gcm_ctx);
+    }
+
+    if (rc != 0) {
+        TRACE_ERROR("ica_aes_gcm_kma_update failed, rc=%ld.\n", rc);
+        rc = CKR_FUNCTION_FAILED;
+        goto done;
+    }
+
+    (*out_data_len) = out_len;
+
+    /* AAD only processed in first update sequence,
+     * mark it empty for all subsequent calls
+     */
+    context->ulAlen = 0;
+
+done:
+    if (buffer)
+        free(buffer);
+
+    return rc;
+}
+
+CK_RV new_gcm_specific_aes_gcm_final(STDLL_TokData_t *tokdata, SESSION *sess,
+                                     ENCR_DECR_CONTEXT *ctx, CK_BYTE *out_data,
+                                     CK_ULONG *out_data_len, CK_BYTE encrypt)
+{
+    CK_RV rc = CKR_OK;
+    AES_GCM_CONTEXT *context = NULL;
+    CK_GCM_PARAMS *aes_gcm_param = NULL;
+    kma_ctx *gcm_ctx = NULL;
+    CK_BYTE *final_tag_data, *buffer = NULL;;
+    CK_ULONG tag_data_len;
+    CK_BYTE tmp_tag[16];
+
+    UNUSED(tokdata);
+    UNUSED(sess);
+    UNUSED(encrypt);
+
+    context = (AES_GCM_CONTEXT *)ctx->context;
+    gcm_ctx = (kma_ctx *)context->ulClen;
+    aes_gcm_param = (CK_GCM_PARAMS *)ctx->mech.pParameter;
+    tag_data_len = (aes_gcm_param->ulTagBits + 7) / 8;
+
+    if (encrypt) {
+        if (context->len != 0) {
+            buffer = (CK_BYTE *)malloc(AES_BLOCK_SIZE);
+            if (!buffer) {
+                TRACE_ERROR("%s\n", ock_err(ERR_HOST_MEMORY));
+                rc = CKR_HOST_MEMORY;
+                goto done;
+            }
+            memcpy(buffer, context->data, context->len);
```

Comment:
Why copying `context->data` to the freshly allocated buffer? Why don't you directly pass `context->data` to `ica_aes_gcm_kma_update` below?

Reply:
done.